### DISCUSSION TITLE
GraphSAINTNodeSampler with CUDA

### DISCUSSION THREAD

**Discussion by stefdesabbata** [pyg-team/pytorch_geometric#7024] [Q&A] [answered]
I am trying to run a graph autoencoder using the `GraphSAINTNodeSampler`. The code works fine on CPU but not on CUDA due to the assertion below in the constructor of `GraphSAINTSampler`, which returns an `AssertionError`.

https://github.com/pyg-team/pytorch_geometric/blob/55fe439334cdc33b03c39fa5ef19429638c54dea/torch_geometric/loader/graph_saint.py#L56

I was wondering why the assertion is there and if there is a way to use `GraphSAINTNodeSampler` on CUDA (also considering that `NeighborLoader` doesn't work either as it returns `NotImplementedError: Could not run 'pyg::neighbor_sample' with arguments from the 'CUDA' backend`).

Thanks in advance for any help and suggestion. :blush:

**Comment by stefdesabbata** [comment, marked as answer]
My question above seems to be the same as the one discussed in [this issue on `GraphSAINTRandomWalkSampler`](https://github.com/pyg-team/pytorch_geometric/issues/2169), which, however, doesn't outline an example solution. I guess the sampler needs to be instantiated using as input a graph dataset on CPU, and then the device conversion (e.g., `to(device)`) needs to be applied to each mini-batch.

For instance

```
loader = GraphSAINTNodeSampler(
        graph,
        batch_size=128,
        num_steps=100,
        sample_coverage=100
    )
```

and then

```
for mini_batch in loader:
    mini_batch = mini_batch.to(device)
```

Would that be the intended usage?

**Reply by rusty1s** [reply]
Yes, this is the intended usage.
